*print-namespace-maps* support

ikappaki opened the pull request:
Hi,

could you please review draft support for the *print-namespace-maps* dynamic variable. It resolves #882.

I implemented the support at the lowest `obj.py` level, though this required to bring in the `keyword.py` and `symbol.py` modules resulting to cyclic dependencies. Not sure how else to implement it, so any other suggestions are most welcome.

I've also blindly followed the example of *print-dup* on how I ended setup the variable in the cli/runtime/object .py files.

In addition, I've used a set to compare the printed output of the map result with multiple k/v, exhaustively covering the possible orderings, is there perhaps a better simpler way to do this? This currently is prune to user errors.

Thanks

chrisrink10 commented on src/basilisp/lang/obj.py:
```diff
@@ -76,17 +77,46 @@ def map_lrepr(
+    elements.
+
+    """
+    from basilisp.lang import keyword as kw  # pylint: disable=cyclic-import
+    from basilisp.lang import symbol as sym  # pylint: disable=cyclic-import
```
I believe the `INamed` interface I created in #884 should suffice to resolve the need for these circular imports.

ikappaki replied:
Hi, and thanks for looking into this.

could you please elaborate a bit how to use the new `INamed` interface.  `basilisp.lang.interfaces` imports `obj.py` and as such if I try to use the same logic there's still a circular dependency (down from two) by  importing `interfaces.py` inline from `object.py`.

Furthermore, with the current logic I need to strip the ns from the symbol or keyword instance, by creating a new instance without it. If I only have the `INamed` interface at hand, i don't think I am able to create an instance of the same type ? for example

currently
``` python
    def entry_reprs():
        for k, v in entries():
            key = k
            if key_ns_drop:
                if isinstance(k, kw.Keyword):
                    key = kw.Keyword(k.name)
                elif isinstance(k, sym.Symbol):
                    key = sym.Symbol(k.name)
            yield f"{lrepr(key, **kwargs)} {lrepr(v, **kwargs)}"
```

with INamed interface
``` python
    def entry_reprs():
        for k, v in entries():
            key = k
            if key_ns_drop:
                if isinstance(k, INamed):
                    key = ?
            yield f"{lrepr(key, **kwargs)} {lrepr(v, **kwargs)}"
```

thanks

chrisrink10 replied:
I think the change in #886 should support this most recent issue:

```python
    def entry_reprs():
        for k, v in entries():
            key = k
            if key_ns_drop:
                if isinstance(k, INamed):
                    key = k.with_name(k.name)
            yield f"{lrepr(key, **kwargs)} {lrepr(v, **kwargs)}"
```

As to the circular dependency, you should be able to move `map_lrepr` over to `src/basilisp/lang/map.py` since that is the only place it is used.

chrisrink10 commented on src/basilisp/lang/obj.py:
```diff
@@ -76,17 +77,46 @@ def map_lrepr(
 
+    def check_same_ns():
+        nses = set()
+        for k, _ in entries():
```
Can you just double check that calling `entries` twice is ok? I think it calls `.items()` which should be fine, but just want to verify we're not exhausting the collection by iterating over it twice.

ikappaki replied:
I can confirm the two instances calling this function are passing in `items`

``` python
@lrepr.register(dict)
def _lrepr_py_dict(o: dict, **kwargs: Unpack[PrintSettings]) -> str:
    return f"#py {map_lrepr(o.items, '{', '}', **kwargs)}"

# and
def _lrepr(self, **kwargs: Unpack[PrintSettings]):
    return map_lrepr(
        self._inner.items, start="{", end="}", meta=self._meta, **kwargs
        )
```

chrisrink10 commented on src/basilisp/lang/obj.py:
```diff
@@ -105,7 +135,7 @@ def entry_reprs():
         return f"^{lrepr(meta, **kwargs)} {start}{seq_lrepr}{end}"
 
-    return f"{start}{seq_lrepr}{end}"
+    return f"{'#:'+ns_same if ns_same else ''}{start}{seq_lrepr}{end}"
```
```suggestion
    return f"{('#:'+ns_same) if ns_same else ''}{start}{seq_lrepr}{end}"
```

This looks slightly less ambiguous to me.

ikappaki replied:
Updated

chrisrink10 commented on src/basilisp/lang/obj.py:
```diff
@@ -76,17 +77,46 @@ def map_lrepr(
+                    break
+        return next(iter(nses)) if len(nses) == 1 else None
+
+    print_namespace_maps = kwargs["print_namespace_maps"]
```
Take note that in #885 I also added typing to these keyword arguments so this will need to be added to the TypedDict definition added there.

ikappaki replied:
Added to `PrintSettings`

``` clojure
class PrintSettings(TypedDict, total=False):
    human_readable: bool
    print_dup: bool
    print_length: PrintCountSetting
    print_level: PrintCountSetting
    print_meta: bool
    print_namespace_maps: bool
    print_readably: bool
```

chrisrink10 reviewed (commented):
Had a few comments. I believe your open question should be resolved as of #885 merging.

chrisrink10 commented on src/basilisp/lang/map.py:
```diff
@@ -107,6 +125,78 @@ def to_persistent(self) -> "PersistentMap[K, V]":
+    def check_same_ns():
+        nses = set()
+        for k, _ in entries():
+            if isinstance(k, INamed):
```
```suggestion
            if isinstance(k, INamed) and k.ns is not None:
```

ikappaki replied:
We actually need the "None" value to go into the `nses` set, to distinguish the case where we have a namespaced key and a non-namespaced key, in which case we don't want to proceed with lifting the key namespace outside the map.

chrisrink10 commented on src/basilisp/lang/map.py:
```diff
@@ -107,6 +125,78 @@ def to_persistent(self) -> "PersistentMap[K, V]":
+    def entry_reprs():
+        for k, v in entries():
+            key = k
+            if key_ns_drop:
+                if isinstance(k, INamed):
+                    key = k.with_name(k.name)
```
```suggestion
            if key_ns_drop and isinstance(k, INamed):
                key = k.with_name(k.name)
```

ikappaki replied:
The `if key_ns_drop` body has been expanded with a new class definition thus I don't think this comment is relevant any more.

chrisrink10 commented on src/basilisp/lang/map.py:
```diff
@@ -107,6 +125,78 @@ def to_persistent(self) -> "PersistentMap[K, V]":
+
+    seq_lrepr = PRINT_SEPARATOR.join(items + trailer)
+
+    print_meta = kwargs["print_meta"]
+    if print_meta and meta:
+        return f"^{lrepr(meta, **kwargs)} {start}{seq_lrepr}{end}"
+
+    return f"{('#:'+ns_same) if ns_same else ''}{start}{seq_lrepr}{end}"
```
```suggestion
    ns_prefix = ("#:" + ns_same) if ns_same else ""
    if kwargs["print_meta"] and meta:
        return f"^{lrepr(meta, **kwargs)} {ns_prefix}{start}{seq_lrepr}{end}"

    return f"{ns_prefix}{start}{seq_lrepr}{end}"
```

This covers the `print_meta` case.

ikappaki replied:
Updated, thanks

chrisrink10 commented on src/basilisp/lang/obj.py:
```diff
@@ -41,7 +43,7 @@ def _dec_print_level(lvl: PrintCountSetting) -> PrintCountSetting:
 
 
-def _process_kwargs(**kwargs: Unpack[PrintSettings]) -> PrintSettings:
+def process_kwargs(**kwargs: Unpack[PrintSettings]) -> PrintSettings:
```
```suggestion
def process_lrepr_kwargs(**kwargs: Unpack[PrintSettings]) -> PrintSettings:
```

Since this is exported, the meaning of the name is a bit more important. I hope this will be a little more explicit.

ikappaki replied:
I've reverted this change to the original _process_kwargs name, because with the new split of the map_lrepr into two, there is no need to expose this fn outside of obj.py

chrisrink10 commented on tests/basilisp/test_core_fns.lpy:
```diff
@@ -2156,7 +2156,19 @@
+    (is (= '(1) (binding [pr (fn [& more] more)] (pr 1)))))
+
+  (testing "with *print-namespace-maps*"
+    (are [res m b] (contains? res (binding [*print-namespace-maps* b] (pr-str m)))
```
Can you add some cases with `*print-meta*` since that's a separate branch.

ikappaki replied:
Tests added, thanks

chrisrink10 commented on src/basilisp/lang/map.py:
```diff
@@ -107,6 +125,78 @@ def to_persistent(self) -> "PersistentMap[K, V]":
+    entries: Callable[[], Iterable[Tuple[Any, Any]]],
+    start: str,
+    end: str,
+    meta=None,
```
```suggestion
    meta: Optional[IPersistentMap] = None,
```

I think this was too hard to type correctly in the previous location, but now it should be fine.

ikappaki replied:
updated, thanks

chrisrink10 commented on src/basilisp/lang/map.py:
```diff
@@ -107,6 +125,78 @@ def to_persistent(self) -> "PersistentMap[K, V]":
+
+    kwargs = process_kwargs(**kwargs)
+
+    def check_same_ns():
```
```suggestion
    def check_same_ns() -> Optional[str]:
```

ikappaki replied:
Updated, thanks

chrisrink10 commented on src/basilisp/lang/map.py:
```diff
@@ -107,6 +125,78 @@ def to_persistent(self) -> "PersistentMap[K, V]":
+    return f"{('#:'+ns_same) if ns_same else ''}{start}{seq_lrepr}{end}"
+
+
+@lrepr.register(dict)
+def _lrepr_py_dict(o: dict, **kwargs: Unpack[PrintSettings]) -> str:
+    return f"#py {map_lrepr(o.items, '{', '}', **kwargs)}"
```
I don't like that this is here since it seems out of place. I'm not sure what to do about that right now though.

ikappaki replied:
Reverted back to the previous scheme

chrisrink10 commented on src/basilisp/lang/map.py:
```diff
@@ -107,6 +117,66 @@ def to_persistent(self) -> "PersistentMap[K, V]":
+    entries_updated = entries
+    if ns_name_shared:
+
+        class EntriesNsRemoved:
+            def __call__(self):
+                for k, v in entries():
+                    yield (k.with_name(k.name), v)
+
+        entries_updated = EntriesNsRemoved()
```
Could this class just be a closure?

ikappaki replied:
Converted into a fn as suggested, thanks

chrisrink10 reviewed (commented):
I appreciate your attempting to resolve the circular dependency issue, but my concern is that we now have duplicated logic for displaying map Lisp representations in two places and my feeling is that is far more likely to cause issues than the circular import.

ikappaki commented:
> I appreciate your attempting to resolve the circular dependency issue, but my concern is that we now have duplicated logic for displaying map Lisp representations in two places and my feeling is that is far more likely to cause issues than the circular import.

Reverted back to the earlier commit, and addressed all of the points made so far. I have to introduce a `py_map` option to `map_lrepr` fn to prevent python dictionary being printed with namespace when that option is on.

chrisrink10 commented:
> > I appreciate your attempting to resolve the circular dependency issue, but my concern is that we now have duplicated logic for displaying map Lisp representations in two places and my feeling is that is far more likely to cause issues than the circular import.
> 
> Reverted back to the earlier commit, and addressed all of the points made so far. I have to introduce a `py_map` option to `map_lrepr` fn to prevent python dictionary being printed with namespace when that option is on.

I initially had this reaction, but I'm not sure I think there's any reason to prevent python dicts from being printed using the same logic? They can be read with a namespace prefix:

```clojure
basilisp.user=> #py #:abc{:def 1 "c" 3}
#py {"c" 3 :abc/def 1}
```

chrisrink10 commented:
Can you rebase against the other PR? I will approve afterwards.

ikappaki commented:
> Can you rebase against the other PR? I will approve afterwards.

Done
